libraries/Wire/src/utility/twi.c

```diff
@@ -396,6 +424,40 @@ void twi_releaseBus(void)
   twi_state = TWI_READY;
 }
 
+/* 
+ * Function twi_setTimeoutInMicros
+ * Desc     set a global timeout for while loops that we might get stuck in
+ * Input    timeout value in microseconds
+ * Output   none
+ */
+void twi_setTimeoutInMicros(uint32_t timeout)
+{
+  twi_timeout_us = timeout;
+}
+
+/* 
+ * Function twi_handleTimeout
+ * Desc     do this stuff when a while loop here has run for longer than twi_timeout_us microseconds
+ * Input    none
+ * Output   none
+ */
+void twi_handleTimeout(void)
+{
+  // remember the bitrate register value
+  uint8_t previous_TWBR = TWBR;
+
+  // remember the address register value
+  uint8_t previous_TWAR = TWBR;
+
+  // reset the interface
+  twi_disable();
+  twi_init();
```

Comment:
I wonder if disable/init is the best approach here. In addition to resetting the TWI module, this also briefly disables the pullups and requires remembering values. Why not just clear and reinitialize TWCR and twiState? In [my version](https://github.com/3devo/ArduinoLibraryWire/blob/f79d0dea0b9623de0d83823c1dc257f1ffb201ef/src/utility/twi.c#L98-L111) I cleared `TWCR` and then called `twi_releaseBus()` to re-initialize it (which just happened to set the correct values, but maybe just copying the contents of it into `twiHandleTimeout` might be cleaner). We should also review whether the other variables (e.g. twi_inRepStart) need to be cleared too. Might very well be that this is not needed when `twi_state` is reset, but would need to check that.

Reply:
It might be a bit heavy handed to disable/reset, yeah, but it feels safer/more universal to me (and maybe even not doing the reset like this is actually why your solution didn't end up solving my lockups). As for this diabling the internal pullups for a moment, I assume that doesn't matter because there must always be external ones on the bus, right? Remembering/resetting the only two values the API lets the user set (frequency & address) seems like no big deal.

I'd be keen to get the PR merged with the reset like this for now and then test with more elegant/faster solutions later, because the only drawback I see in doing it this way is that it might be slightly slower than another solution with the main advantage being that it's a proper reset. And maybe honestly an even more robust solution here would be to put some hard-coded delay in to let the slaves' state-machines time out to guarantee everyone is all back in sync when the reset procedure is completed.

Reply:
> As for this diabling the internal pullups for a moment, I assume that doesn't matter because there must always be external ones on the bus, right? 

Well, if there were, why do we even enable the internal pullups? I guess for the cases where people just attach something without worrying about this. But yeah, it's not terrible.

Overall, agreed that this is not critical, I expect your current code to work.

Reply:
I'll click the resolve button on this thread then :+1:

Reply:
Maybe it would be good solution to **not to handle the timeout** at all in the library? Instead of this set some kind of flag that will indicate the timeout. The developer can check this flag in the **loop** method. In case of timeout, he can do some additional things there, like:
 * clear the timeout flag (new method for the Wire library required)
 * reset TWI interface by himself
 * possible reset some other slave TWI devices (about these, the TWI library knows nothing) 

It was also already mentioned here: https://github.com/wmarkow/Arduino/commit/afc25a64316e38f29cf887f546b8a09c73375838#comments

Reply:
Not resetting on timeout is indeed an interesting thought. That would also be helpful in true multi-master situations. Rather than setting a very long timeout, to allow waiting for the other master to be done, you could set a short timeout and let your sketch do something else while the other master is busy, and then retry the I2c transaction after a while.

However, I think that for most usecases, just "reset and make it work again" is the most commonly required behaviour, so it would be cumbersome if you are obliged to always implement this in a sketch (even more tricky when libraries use Wire).

How about making the reset optional? e.g. by calling `setWireTimeout(100, WIRE_TIMEOUT_RESET)` or `setWireTimeout(100, WIRE_TIMEOUT_REPORT)`, probably defaulting to reset?

Reply:
>However, I think that for most usecases, just "reset and make it work again" is the most commonly required behaviour, so it would be cumbersome if you are obliged to always implement this in a sketch (even more tricky when libraries use Wire).

Agreed there. How about we set a flag whenever we enter the timeout handler and leave it to the user to clear that flag? That way they always have a way to know if they had a timeout, no matter if we do the reset or let them handle it. Then we make the API `setWireTimeoutUs(uint32_t timeout, bool reset_with_timeout)` and they can choose to if they want the library to reset on timeout or not.

So the Wire library would get three new functions:
```
bool readResetFlag(void);
void clear(void);
void setWireTimeoutUs(uint32_t timeout, bool reset_with_timeout)
```

Reply:
Sounds like a good plan. I do think that "ResetFlag" is not the right name, since no reset might have happened. Something like "TimeoutOccurred"? Also, "clear" is way too general, that should probably be more specific.

We could consider making one function instead of two, e.g. `getAndClearTimeoutOccurredFlag()`, since typically you would do both at the same time. OTOH, I can image a library that wants to know the flag, but leave it set for the sketch, so maybe two functions is actually just fine. Maybe `getTimeoutOccurred()` and `clearTimeoutOccurred()`?

Reply:
Oops didn't mean to name the funccion simply `clear` (meant to put `clearResetFlag` there, fixed now.)

Reply:
I can combine the functions.

@wmarkow How would that work for you?
```
bool getWireTimeoutOccurred(bool clear_after_read);
void setWireTimeoutUs(uint32_t timeout, bool reset_with_timeout);
```
So if you want to handle the timeout yourself you can do
```c
setWireTimeoutUs(10000, false);

void loop() {
  if ( getWireTimeoutOccurred(false) ){
    // do your manual timeout handling stuff here
    getWireTimeoutOccurred(true);  // reset the flag
  }
}
```
But if you want the auto-reset then you just setup the timeout like this:
```c
setWireTimeoutUs(10000, true);
```
And you don't have to bother with the polling and resetting yourself.

Reply:
Good, that the case is moving forward.
I would use a **flag** word when it goes about timeout event. How about three new methods:
```c++
/***
 * Sets the TWI timeout.
 *
 * @param timeout a timeout in microseconds
 * @param reset_with_timeout if true then TWI interface will be automatically reset on timeout
 *                           if false then TWI interface will not be reset on timeout
 */
void setWireTimeoutUs(uint32_t timeout, bool reset_with_timeout);

/***
 * Returns the TWI timeout flag.
 *
 * @return true if timeout occured
 */
bool getWireTimeoutFlag();

/***
 * Clears the TWI timeout flag.
 */
void clearWireTimeoutFlag();
```

or 2 new methods:
```c++
/***
 * Sets the TWI timeout.
 *
 * @param timeout a timeout in microseconds
 * @param reset_with_timeout if true then TWI interface will be automatically reset on timeout
 */
void setWireTimeoutUs(uint32_t timeout, bool reset_with_timeout);

/***
 * Returns the TWI timeout flag and clears the flag.
 *
 * @return true if timeout occured
 */
bool getWireTimeoutFlag();
````
The timeout flag can be cleared automatically when reading it. The main scenario is mostly like @greyltc has proposed above: read the flag, if it is true then execute some additional code and clear the flag. I'm just thinking if there is a use case when the developer wants to read the flag but not to clear it.

Reply:
in da7d85a, I implemented these three user-facing functions to manage timeouts
```
void setWireTimeoutUs(uint32_t timeout, bool reset_with_timeout);
bool getWireTimeoutFlag();
void clearWireTimeoutFlag();
```
so I'll resolve this now.